### PR TITLE
fhirpath parsing now includes comments and whitespace into the Expression object model.

### PR DESCRIPTION
**Pull request by brianpos**
## Description
This extends on the work in the fhirpath parsing changes to be able to capture the whitespace including comments to permit exact round trips with the Expression tree.

## Related issues
#2517 

## Testing
There are many unit tests for this, and also tested with my fhirpath validator project locally to verify that it sucessfully roundtripped all the expressions in the R4B/R5 invariants, searchparameters and about 10 large implementation guides.

## FirelyTeam Checklist
- [ ] **Update the title** of the PR to be succinct and less than 50 characters
- [ ] Mark the PR with the label **breaking change** when this PR introduces breaking changes

### CODE REVIEW

**Comment by brianpos** [comment]
I expect this will be just reviewed in the diffs, then pushed direct into the other branch before (or after) going to develop.

File: src/Hl7.Fhir.Base/FhirPath/Parser/Grammar.cs [hunk 1]
```diff
@@ -165,36 +200,55 @@ public static Parser<Expression> Invocation(Expression focus)
...
         public static Parser<Expression> IndexerInvocation(Expression focus)
         {
-            return Parse.Contained(Expression, Parse.Char('[').TokenIgnoreComments(), Parse.Char(']').TokenIgnoreComments())
-                .Select(ix => new IndexerExpression(focus, ix)).Positioned();
+            return
+                // The whitespace before the '[' is captured in the ws1 token
+                // so that the position information for the indexer isn't confused with the comment
+                from wsLeading in WhitespaceOrComments()
+                from indexer in (
+                    from lparen in Parse.Char('[').Select(v => new SubToken(v)).Positioned()
```
Comment (ewoutkramer) [inline src/Hl7.Fhir.Base/FhirPath/Parser/Grammar.cs:215]: Where is the equivalent of the `TokenIgnoreComments` on the left hand side? So, where are ww capturing the whitespace between the `[` and the following expression?  Or is Expression capturing those?

File: src/Hl7.Fhir.Base/FhirPath/Parser/Lexer.cs [hunk 2]
```diff
@@ -68,7 +68,7 @@ internal partial class Lexer
 
         public static readonly Parser<string> DelimitedIdentifier =
 
-            DelimitedContents('"')
+            DelimitedContents('"') // this isn't to standard
```
Comment (ewoutkramer) [inline src/Hl7.Fhir.Base/FhirPath/Parser/Lexer.cs]: It was there for backwards compat. Should we axe it?

**Review by ewoutkramer — DISMISSED** [review]

File: src/Hl7.Fhir.Base/FhirPath/Parser/Lexer.cs (hunk 2 above)
Comment (brianpos) [reply src/Hl7.Fhir.Base/FhirPath/Parser/Lexer.cs]: I'm OK with it remaining for that purpose. Maybe I'll just update the code comment to say that's why it's there?

File: src/Hl7.Fhir.Base/FhirPath/Parser/Grammar.cs (hunk 1 above)
Comment (brianpos) [reply src/Hl7.Fhir.Base/FhirPath/Parser/Grammar.cs:215]: Expression captures those - just as it does in function arguments between the commas.
There is a unit test that covers this specific case too.
https://github.com/FirelyTeam/firely-net-sdk/blob/6046024baf75120465423899e9c343ede45c4e97/src/Hl7.FhirPath.Tests/Tests/FhirPathGrammarTest.cs#L372

**Review by Kasdejong — APPROVED** [review]

**Review by ewoutkramer — APPROVED** [review]

**Pull request merged**